### ISSUE TITLE
null ptr deref, segfault in S_rv2gv (pp.c:296)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#15607]
Migrated from [rt.perl.org#129288](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=129288) (status was 'open')

Searchable as RT129288$

**Comment by p5pRT** [comment]
### From @geeknik

Triggered in v5\.25\.5 \(v5\.25\.4\-130\-g7aa7bbc\) with AFL \+ ASAN \+ libdislocator\.

\./perl \-e 'select$0=%&#8203;::=\*0==0'
ASAN&#8203;:SIGSEGV
=================================================================
==3754==ERROR&#8203;: AddressSanitizer&#8203;: SEGV on unknown address 0x00000000000f \(pc 0x0000009a4fb9 bp 0x621000010f2c sp 0x7ffe58dc62b0 T0\)
&nbsp;    \#0 0x9a4fb8 in S\_rv2gv /root/perl/pp\.c&#8203;:296&#8203;:6
&nbsp;    \#1 0x9a419a in Perl\_pp\_rv2gv /root/perl/pp\.c&#8203;:312&#8203;:10
&nbsp;    \#2 0x7f4583 in Perl\_runops\_debug /root/perl/dump\.c&#8203;:2239&#8203;:23
&nbsp;    \#3 0x5a11c6 in S\_run\_body /root/perl/perl\.c&#8203;:2525&#8203;:2
&nbsp;    \#4 0x5a11c6 in perl\_run /root/perl/perl\.c&#8203;:2448
&nbsp;    \#5 0x4de5fd in main /root/perl/perlmain\.c&#8203;:123&#8203;:9
&nbsp;    \#6 0x7f375e120b44 in \_\_libc\_start\_main /build/glibc\-uPj9cH/glibc\-2\.19/csu/libc\-start\.c&#8203;:287
&nbsp;    \#7 0x4de26c in \_start \(/root/perl/perl\+0x4de26c\)

AddressSanitizer can not provide additional info\.
SUMMARY&#8203;: AddressSanitizer&#8203;: SEGV /root/perl/pp\.c&#8203;:296 S\_rv2gv
==3754==ABORTING

**Comment by p5pRT** [comment]
### From zefram@fysh.org

The test case no longer segvs for me\, since v5\.25\.6\-78\-g8b0c337\.

\-zefram

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, Jan 27\, 2017 at 06&#8203;:52&#8203;:50AM \+0000\, Zefram wrote&#8203;:
> The test case no longer segvs for me\, since v5\.25\.6\-78\-g8b0c337\.

The code can be reduced to&#8203;:

&nbsp;    $x = "foo";
&nbsp;    %&#8203;:: = \($x\);
&nbsp;    \*$x;

prior to that commit\, pp\_assign when assigning to a hash with an odd
number of elements on the RHS\, ended up using PL\_sv\_undef as the hash
value rather than a copy of PL\_sv\_undef\. PL\_sv\_undef as a value in a stash
apparently means something special\, so the glob lookup\, \*$x\, crashed\.

That commit fixed the bug of storing PL\_sv\_undef; I'm not sure whether
there's still a bug of \*$x crashing if $&#8203;::\{$x\} is a pointer to
PL\_sv\_undef \- whether that is legal value etc\.

Basically the gv\_fetchsv\_nomg\(\) call in S\_rv2gv\(\) returns NULL\, which
S\_rv2gv\(\) isn't expecting\.

Perl\_gv\_fetchpvn\_flags\(\) has&#8203;:

&nbsp;    gvp = \(GV\*\*\)hv\_fetch\(stash\,name\,is\_utf8 ? \-\(I32\)len : \(I32\)len\,add\);
&nbsp;    if \(\!gvp || \*gvp == \(const GV \*\)&PL\_sv\_undef\) \{
&nbsp;    if \(addmg\) gv = \(GV \*\)newSV\(0\);
&nbsp;    else return NULL;

which is where the test for the magic value PL\_sv\_undef appears\.

\-\- 
Never work with children\, animals\, or actors\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue\, 28 Mar 2017 08&#8203;:33&#8203;:20 \-0700\, davem wrote&#8203;:
> On Fri\, Jan 27\, 2017 at 06&#8203;:52&#8203;:50AM \+0000\, Zefram wrote&#8203;:
> > The test case no longer segvs for me\, since v5\.25\.6\-78\-g8b0c337\.
> 
> The code can be reduced to&#8203;:
> 
>     $x = "foo";
>     %&#8203;:: = \($x\);
>     \*$x;
> 
> prior to that commit\, pp\_assign when assigning to a hash with an odd
> number of elements on the RHS\, ended up using PL\_sv\_undef as the hash
> value rather than a copy of PL\_sv\_undef\. PL\_sv\_undef as a value in a stash
> apparently means something special\, so the glob lookup\, \*$x\, crashed\.
> 
> That commit fixed the bug of storing PL\_sv\_undef; I'm not sure whether
> there's still a bug of \*$x crashing if $&#8203;::\{$x\} is a pointer to
> PL\_sv\_undef \- whether that is legal value etc\.

Then one can still make it crash&#8203;:

use feature "refaliasing";
\\$&#8203;::\{foo\} = \\undef;
\*\{"foo"\};

$ pbpaste|\./perl \-Ilib
Aliasing via reference is experimental at \- line 2\.
Segmentation fault&#8203;: 11

> Basically the gv\_fetchsv\_nomg\(\) call in S\_rv2gv\(\) returns NULL\, which
> S\_rv2gv\(\) isn't expecting\.
> 
> Perl\_gv\_fetchpvn\_flags\(\) has&#8203;:
> 
>     gvp = \(GV\*\*\)hv\_fetch\(stash\,name\,is\_utf8 ? \-\(I32\)len : \(I32\)len\,add\);
>     if \(\!gvp || \*gvp == \(const GV \*\)&PL\_sv\_undef\) \{
>     if \(addmg\) gv = \(GV \*\)newSV\(0\);
>     else return NULL;
> 
> which is where the test for the magic value PL\_sv\_undef appears\.

And that seems to be ancient\.  However\, I don’t know what would be logical to do otherwise\.

Interestingly\, other immortals are also\, but differently\, problematic\.

use feature "refaliasing";
\\$&#8203;::\{foo\} = \\\!1;
\*\{"foo"\};

Aliasing via reference is experimental at \- line 2\.
Assertion failed&#8203;: \(\!SvPOKp\(gv\)\)\, function Perl\_gv\_init\_pvn\, file gv\.c\, line 411\.
Abort trap&#8203;: 6

Perhaps they should just behave like this instead&#8203;:

use feature "refaliasing";
\\$&#8203;::\{foo\} = \\"foo";
\*\{"foo"\};

$ pbpaste|\./perl \-Ilib
Aliasing via reference is experimental at \- line 2\.
Modification of a read\-only value attempted at \- line 3\.

But the error message there is not all that helpful in saying what went wrong\.

Maybe we could change croak\_no\_modify to include the op description\.  \(How much would that break?\)

\-\- 

Father Chrysostomos
